### PR TITLE
ARROW-9176: [Rust] Fix for memory leaks in Arrow allocator

### PR DESCRIPTION
**Pull request by maxburke**
If realloc is passed a block of size zero, free the block and return the
address of the bypass block.

### CODE REVIEW

File: rust/arrow/src/memory.rs [hunk 1]
```diff
@@ -150,25 +150,32 @@ pub fn allocate_aligned(size: usize) -> *mut u8 {
...
-        if !new_ptr.is_null() && new_size > old_size {
-            new_ptr.add(old_size).write_bytes(0, new_size - old_size);
-        }
-        new_ptr
+        return allocate_aligned(new_size);
+    }
+
+    if new_size == 0 {
+        free_aligned(ptr, old_size);
+        return BYPASS_PTR.as_ptr();
```
Comment (houqp) [inline rust/arrow/src/memory.rs:165]: good catch, would be good to add a test for this.

**Comment by maxburke** [comment]
Seeing some weird issues in long running applications; there are no leaks as reported by valgrind anymore, but memory use is up. I'm going to close this PR for now.

**Comment by maxburke** [comment]
Re-opening this PR because the memory use changes are related to our use of the underlying system allocator, not this change.

File: rust/arrow/src/memory.rs (hunk 1 above)
Comment (maxburke) [reply rust/arrow/src/memory.rs:165]: What kind of test are you looking for? Running valgrind on the tests to validate there are no leaked blocks?

Comment (houqp) [reply rust/arrow/src/memory.rs:165]: At first i thought there is straightforward api from the global allocator we can use to access memory allocation metadata. This turned out not to be the case, which means we will either have to introduce valgrind or change rust global allocator to a custom one for tests. Both approaches will expand the scope of this PR too much, so scratch what I said, I think it's better to be done as a separate PRs.

Comment (paddyhoran) [reply rust/arrow/src/memory.rs:165]: I guess we could start running valgrind in CI for Rust?  I don't use Valgrind but I created an [issue](https://issues.apache.org/jira/browse/ARROW-9200) if anyone wants to pick it up.

**Review by paddyhoran — APPROVED** [review]

**Pull request closed without merging**